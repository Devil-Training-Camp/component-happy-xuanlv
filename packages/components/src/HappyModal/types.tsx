import type React from 'react';
import type { HappyButtonProps } from '../HappyButton';
import type { HappyDialogProps } from '../HappyDialog/types';

/**
 * 弹窗参数
 */
export interface HappyModalProps extends HappyDialogProps {
  /**
   * Modal body 样式
   */
  bodyStyle?: React.CSSProperties;
  /**
   * 取消按钮文字
   * @defaultValue '取消'
   */
  cancelText?: React.ReactNode;
  /**
   * 是否显示右上角的关闭按钮
   * @defaultValue true
   */
  closable?: boolean;
  /**
   * 确定按钮 loading
   */
  confirmLoading?: boolean;
  /**
   * 底部内容，当不需要默认底部按钮时，可以设为 footer={null}
   */
  footer?: React.ReactNode | ((config: RenderFooterConfig) => React.ReactNode);
  /**
   * 点击确定回调
   */
  onOk?: (e: React.MouseEvent<HTMLElement>) => void;
  /**
   * 确认按钮文字
   * @defaultValue '确定'
   */
  okText?: React.ReactNode;
  /**
   * 标题
   */
  title?: React.ReactNode;
  /**
   * ok 按钮 props
   */
  okButtonProps?: HappyButtonProps;
}

interface RenderFooterConfig {
  /**
   * 取消节点
   */
  cancelButton: React.ReactNode;
  /**
   * 确认节点
   */
  okButton: React.ReactNode;
}
