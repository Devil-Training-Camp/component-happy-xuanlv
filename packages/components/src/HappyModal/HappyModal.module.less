@import '../style/variable-default';

.modalHeader {
  padding: @happy-modal-header-padding;
  color: @happy-text-color;
  background: @happy-modal-header-bg;
  border-bottom: @happy-modal-header-border-width @happy-modal-header-border-style
    @happy-modal-header-border-color-split;
  border-radius: @happy-border-radius-base @happy-border-radius-base 0 0;
}

.modalTitle {
  margin: 0;
  font-size: @happy-modal-header-title-font-size;
  font-weight: bold;
  line-height: @happy-modal-header-title-line-height;
  color: @happy-modal-heading-color;
  word-wrap: break-word;
}

.modalContent {
  padding: @happy-modal-body-padding;
  font-size: @happy-font-size-base;
  line-height: @happy-line-height-base;
  word-wrap: break-word;
}

.modalFooter {
  padding: @happy-modal-footer-padding-vertical @happy-modal-footer-padding-horizontal;
  text-align: right;
  background: @happy-modal-footer-bg;
  border-top: @happy-modal-footer-border-width @happy-modal-footer-border-style
    @happy-modal-footer-border-color-split;
  border-radius: 0 0 @happy-border-radius-base @happy-border-radius-base;
}

.footerButton {
  margin-left: 8px;
}
