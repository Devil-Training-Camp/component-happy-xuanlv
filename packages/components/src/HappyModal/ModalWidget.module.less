@import '../style/variable-default';

.closeButton {
  position: absolute;
  top: 0;
  right: 0;
  box-sizing: content-box;
  display: flex;
  align-items: center;
  justify-content: center;
  width: @happy-modal-close-size;
  height: @happy-modal-close-size;
  color: @happy-modal-close-color;
  cursor: pointer;
  transition: color 0.2s ease-out;

  &:hover,
  &:focus {
    color: @happy-icon-color-hover;
  }
}

.closeIcon {
  font-size: @happy-font-size-lg;
  fill: currentColor;
}

.footerButton {
  margin-left: 8px;
}
