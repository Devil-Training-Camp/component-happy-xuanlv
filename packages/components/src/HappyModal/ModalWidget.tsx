import React from 'react';
import type { HappyButtonProps } from '../HappyButton';
import { HappyButton } from '../HappyButton';
import styles from './ModalWidget.module.less';

export interface CloseButtonProps {
  onClick?: React.EventHandler<React.MouseEvent>;
}

/**
 * 关闭按钮
 */
export function CloseButton({ onClick }: CloseButtonProps) {
  return (
    <span className={styles.closeButton} onClick={onClick}>
      <svg
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
        width="1em"
        height="1em"
        viewBox="0 0 32 32"
        className={styles.closeIcon}
      >
        <path d="M17.62 16l8.2-9.78c0.036-0.043 0.059-0.099 0.059-0.16 0-0.136-0.11-0.247-0.246-0.248h-2.494c-0.154 0.002-0.291 0.071-0.383 0.179l-0.001 0.001-6.766 8.062-6.768-8.064c-0.092-0.109-0.229-0.178-0.383-0.178-0.001 0-0.001 0-0.002 0h-2.492c-0.001 0-0.001 0-0.002 0-0.137 0-0.248 0.111-0.248 0.248 0 0.062 0.023 0.119 0.061 0.162l-0-0 8.202 9.778-8.202 9.78c-0.036 0.043-0.059 0.099-0.059 0.16 0 0.137 0.111 0.248 0.248 0.248 0 0 0 0 0.001 0h2.494c0.154-0.002 0.291-0.071 0.383-0.179l0.001-0.001 6.766-8.064 6.766 8.066c0.092 0.109 0.229 0.178 0.383 0.178 0.001 0 0.001 0 0.002 0h2.494c0.001 0 0.001 0 0.002 0 0.137 0 0.248-0.111 0.248-0.248 0-0.062-0.023-0.119-0.061-0.162l0 0-8.204-9.778z" />
      </svg>
    </span>
  );
}

export interface FooterOkButtonProps {
  confirmLoading?: boolean;
  okText?: React.ReactNode;
  onOk?: React.EventHandler<React.MouseEvent>;
  okButtonProps?: HappyButtonProps;
}

/**
 * 弹窗底部确认按钮
 */
export function FooterOkButton({
  confirmLoading,
  okText,
  onOk,
  okButtonProps,
}: FooterOkButtonProps) {
  return (
    <HappyButton
      className={styles.footerButton}
      type="primary"
      onClick={onOk}
      loading={confirmLoading}
      {...okButtonProps}
    >
      {okText ?? '确定'}
    </HappyButton>
  );
}

export interface FooterCancelButtonProps {
  cancelText?: React.ReactNode;
  onCancel?: React.EventHandler<React.MouseEvent>;
}

/**
 * 弹窗底部取消按钮
 */
export function FooterCancelButton({ cancelText, onCancel }: FooterCancelButtonProps) {
  return (
    <HappyButton className={styles.footerButton} onClick={onCancel}>
      {cancelText ?? '取消'}
    </HappyButton>
  );
}

export function IconInfoCircle(props: React.HTMLAttributes<HTMLOrSVGElement>) {
  return (
    <svg
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
      width="1em"
      height="1em"
      viewBox="0 0 32 32"
      {...props}
    >
      <path d="M16 2c7.732 0 14 6.268 14 14s-6.268 14-14 14v0c-7.732 0-14-6.268-14-14s6.268-14 14-14v0zM16 4.376c-6.327 0.121-11.41 5.279-11.41 11.624s5.083 11.503 11.399 11.624l0.011 0c6.327-0.121 11.41-5.279 11.41-11.624s-5.083-11.503-11.399-11.624l-0.011-0zM16.75 14c0.138 0 0.25 0.112 0.25 0.25v8.5c0 0.138-0.112 0.25-0.25 0.25v0h-1.5c-0.138 0-0.25-0.112-0.25-0.25v0-8.5c0-0.138 0.112-0.25 0.25-0.25zM16 9c0.828 0 1.5 0.672 1.5 1.5s-0.672 1.5-1.5 1.5v0c-0.828 0-1.5-0.672-1.5-1.5s0.672-1.5 1.5-1.5v0z" />
    </svg>
  );
}
