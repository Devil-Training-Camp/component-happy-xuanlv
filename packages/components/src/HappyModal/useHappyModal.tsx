import React from 'react';
import { useMethods } from '@xuan/hooks';
import type { HappyConfirmProps } from '../confirm/types';
import type { HappyModalProps } from './types';
import HappyModal from './HappyModal';
import HappyConfirm from '../confirm/HappyConfirm';
import { happyConfirmBuilder } from '../confirm/confirm';

export interface APIHappyModalProps extends HappyModalProps {
  /**
   * 具体内容
   */
  content?: React.ReactNode;
}

/**
 * 通过 useModal 创建支持读取 context 的 contextHolder
 */
export default function useHappyModal() {
  // 每个弹窗唯一 ID
  const modalIdRef = React.useRef(0);
  // 暂存的节点
  const [holderNodes, setHolderNodes] = React.useState<Map<number, JSX.Element>>(
    new Map(),
  );

  function createBuilder<Config extends HappyConfirmProps>(config: Config, ModalComponent: React.FC) {
    const currentModalKey = (modalIdRef.current += 1);

    return happyConfirmBuilder(config, (currentConfig) => {
      // 更新组件
      if (currentConfig) {
        setHolderNodes((prevState) => {
          return new Map(prevState).set(
            currentModalKey,
            <ModalComponent key={currentModalKey} {...currentConfig} />,
          );
        });
        return;
      }

      // 卸载组件
      setHolderNodes((prevState) => {
        const currentMap = new Map(prevState);
        currentMap.delete(currentModalKey);
        return currentMap;
      });
    });
  }

  const methods = useMethods({
    /**
     * API 调用弹出确认框
     */
    confirm(config: HappyConfirmProps) {
      return createBuilder(config, HappyConfirm);
    },
    /**
     * API 调用弹出 Modal
     */
    open({ content, ...config }: APIHappyModalProps) {
      return createBuilder({ children: content, ...config }, HappyModal);
    },
  });

  return [methods, Array.from(holderNodes.values())] as const;
}
