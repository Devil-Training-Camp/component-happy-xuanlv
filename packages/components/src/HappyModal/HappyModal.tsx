import React from 'react';
import type { HappyModalProps } from './types';
import { HappyDialog } from '../HappyDialog';
import { CloseButton, FooterCancelButton, FooterOkButton } from './ModalWidget';
import styles from './HappyModal.module.less';

/**
 * 弹窗
 */
export default function HappyModal({
  bodyStyle,
  cancelText,
  closable = true,
  confirmLoading,
  footer,
  onCancel,
  onOk,
  okText,
  title,
  okButtonProps,
  children,
  ...baseProps
}: HappyModalProps) {
  function renderFooter() {
    if (footer === null) {
      return null;
    }
    const cancelButton = <FooterCancelButton onCancel={onCancel} cancelText={cancelText} />;
    const okButton = (
      <FooterOkButton
        confirmLoading={confirmLoading}
        onOk={onOk}
        okText={okText}
        okButtonProps={okButtonProps}
      />
    );

    return (
      <div className={styles.modalFooter}>
        {typeof footer === 'function'
          ? footer({ cancelButton, okButton })
          : footer ?? (
              <>
                {cancelButton}
                {okButton}
              </>
            )}
      </div>
    );
  }

  return (
    <HappyDialog onCancel={onCancel} {...baseProps}>
      {title && (
        <div className={styles.modalHeader}>
          <div className={styles.modalTitle}>{title}</div>
        </div>
      )}
      {closable && <CloseButton onClick={onCancel} />}
      <div className={styles.modalContent} style={bodyStyle}>
        {children}
      </div>
      {renderFooter()}
    </HappyDialog>
  );
}
