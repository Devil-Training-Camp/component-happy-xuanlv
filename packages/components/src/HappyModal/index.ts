import HappyModal from './HappyModal';
import useHappyModal from './useHappyModal';

export { HappyModal, useHappyModal };
