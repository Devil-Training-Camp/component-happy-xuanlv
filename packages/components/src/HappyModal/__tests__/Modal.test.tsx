import React, { useEffect, useRef, useState } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render } from '@testing-library/react';
import HappyModal from '../HappyModal';
import type { HappyModalProps } from '../types';
import mountTest from '../../../tests/shared/mountTest';
import snapshotTest from '../../../tests/shared/snapshotTest';

function ModalTester(props: HappyModalProps) {
  const containerRef = useRef<HTMLDivElement>(null);
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    setVisible(true);
  }, []);

  return (
    <>
      <div ref={containerRef} />
      <HappyModal
        getContainer={() => containerRef.current!}
        visible={visible}
        onCancel={() => setVisible(false)}
        {...props}
      >
        Modal Content
      </HappyModal>
    </>
  );
}

describe('HappyModal', () => {
  mountTest(<ModalTester />);

  it('正常渲染', async () => {
    snapshotTest(<ModalTester />);
  });

  //
  it('渲染无 footer', () => {
    snapshotTest(<ModalTester footer={null} />);
  });

  it('onCancel 需要被调用', () => {
    const onCancel = vi.fn();
    render(<HappyModal visible onCancel={onCancel} />);
    fireEvent.click(document.body.querySelectorAll('button')[0]);
    expect(onCancel).toHaveBeenCalled();
  });

  it('onOk 需要被调用', () => {
    const onOk = vi.fn();
    render(<HappyModal visible onOk={onOk} />);
    fireEvent.click(document.body.querySelectorAll('button')[1]);
    expect(onOk).toHaveBeenCalled();
  });

  it('鼠标触发位置', () => {
    const Demo = () => {
      const [open, setOpen] = React.useState(false);
      const containerRef = React.useRef<HTMLDivElement>(null);
      return (
        <div ref={containerRef}>
          <div id="trigger" onClick={() => setOpen(true)}>
            click me
          </div>
          <HappyModal visible={open} getContainer={() => containerRef.current!} />
        </div>
      );
    };
    const { container } = render(<Demo />);
    fireEvent.click(container.querySelectorAll('#trigger')[0]);
    expect(
      (container.querySelectorAll('[role=dialog]')[0] as HTMLDivElement).style.transformOrigin,
    ).toBeTruthy();
  });
});
