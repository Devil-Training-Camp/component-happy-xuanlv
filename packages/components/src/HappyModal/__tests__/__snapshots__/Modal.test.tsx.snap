// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HappyModal > 正常渲染 1`] = `
<div
  style="overflow: hidden;"
>
  <div
    class="_happyModalRoot_c5e8fd"
    tabindex="-1"
  >
    <div
      class="_happyModalMask_c5e8fd _fadeIn_c5e8fd"
    />
    <div
      class="_happyModalWrap_c5e8fd"
    >
      <div
        class="_happyModal_c5e8fd _animateIn_c5e8fd"
        role="dialog"
        style="width: 416px;"
      >
        <span
          class="_closeButton_472604"
        >
          <svg
            class="_closeIcon_472604"
            height="1em"
            version="1.1"
            viewBox="0 0 32 32"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.62 16l8.2-9.78c0.036-0.043 0.059-0.099 0.059-0.16 0-0.136-0.11-0.247-0.246-0.248h-2.494c-0.154 0.002-0.291 0.071-0.383 0.179l-0.001 0.001-6.766 8.062-6.768-8.064c-0.092-0.109-0.229-0.178-0.383-0.178-0.001 0-0.001 0-0.002 0h-2.492c-0.001 0-0.001 0-0.002 0-0.137 0-0.248 0.111-0.248 0.248 0 0.062 0.023 0.119 0.061 0.162l-0-0 8.202 9.778-8.202 9.78c-0.036 0.043-0.059 0.099-0.059 0.16 0 0.137 0.111 0.248 0.248 0.248 0 0 0 0 0.001 0h2.494c0.154-0.002 0.291-0.071 0.383-0.179l0.001-0.001 6.766-8.064 6.766 8.066c0.092 0.109 0.229 0.178 0.383 0.178 0.001 0 0.001 0 0.002 0h2.494c0.001 0 0.001 0 0.002 0 0.137 0 0.248-0.111 0.248-0.248 0-0.062-0.023-0.119-0.061-0.162l0 0-8.204-9.778z"
            />
          </svg>
        </span>
        <div
          class="_modalContent_a1e552"
        >
          Modal Content
        </div>
        <div
          class="_modalFooter_a1e552"
        >
          <button
            class="_happyButton_5fc01e _footerButton_472604"
            type="button"
          >
            取消
          </button>
          <button
            class="_happyButton_5fc01e _footerButton_472604 _primary_5fc01e"
            type="button"
          >
            确定
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HappyModal > 渲染无 footer 1`] = `
<div
  style="overflow: hidden;"
>
  <div
    class="_happyModalRoot_c5e8fd"
    tabindex="-1"
  >
    <div
      class="_happyModalMask_c5e8fd _fadeIn_c5e8fd"
    />
    <div
      class="_happyModalWrap_c5e8fd"
    >
      <div
        class="_happyModal_c5e8fd _animateIn_c5e8fd"
        role="dialog"
        style="width: 416px;"
      >
        <span
          class="_closeButton_472604"
        >
          <svg
            class="_closeIcon_472604"
            height="1em"
            version="1.1"
            viewBox="0 0 32 32"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.62 16l8.2-9.78c0.036-0.043 0.059-0.099 0.059-0.16 0-0.136-0.11-0.247-0.246-0.248h-2.494c-0.154 0.002-0.291 0.071-0.383 0.179l-0.001 0.001-6.766 8.062-6.768-8.064c-0.092-0.109-0.229-0.178-0.383-0.178-0.001 0-0.001 0-0.002 0h-2.492c-0.001 0-0.001 0-0.002 0-0.137 0-0.248 0.111-0.248 0.248 0 0.062 0.023 0.119 0.061 0.162l-0-0 8.202 9.778-8.202 9.78c-0.036 0.043-0.059 0.099-0.059 0.16 0 0.137 0.111 0.248 0.248 0.248 0 0 0 0 0.001 0h2.494c0.154-0.002 0.291-0.071 0.383-0.179l0.001-0.001 6.766-8.064 6.766 8.066c0.092 0.109 0.229 0.178 0.383 0.178 0.001 0 0.001 0 0.002 0h2.494c0.001 0 0.001 0 0.002 0 0.137 0 0.248-0.111 0.248-0.248 0-0.062-0.023-0.119-0.061-0.162l0 0-8.204-9.778z"
            />
          </svg>
        </span>
        <div
          class="_modalContent_a1e552"
        >
          Modal Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
